### ISSUE TITLE
Rails 4.2.1: ActiveRecord's save and save(validate: false) methods issue with serialized attributes

### ISSUE THREAD

**Issue by veera** [rails/rails#19960]
This works fine in Rails 4.1

For example consider a User model with a serialize attribute `options`. `null:false` for this attribute

Schema

```
 create_table :users do |t|
   t.string :options, null: false
   t.timestamps null: false
 end
```

Model

```
class User < ActiveRecord::Base
  serialize :options, Array
end
```

Failing test

```
class UserTest < ActiveSupport::TestCase
  test "when there is no options value with save" do
    user = User.new
    user.save
  end

  test "when there is an empty array options with save" do
    user = User.new
    user.options = []
    user.save
  end

  test "when there is no options value with validate false" do
    user = User.new
    user.save(validate: false)
  end

  test "when there is an empty array options with validate false" do
    user = User.new
    user.options = []
    user.save(validate: false)
  end
end
```

Running above test, getting this error, Also seen the same issue on MySQL.
```
ActiveRecord::StatementInvalid: SQLite3::ConstraintException: NOT NULL constraint failed: products.options: INSERT INTO "users" ("created_at", "updated_at") VALUES (?, ?)
```

It seems even though if i set model attribute(options) value is empty array, it doesn't available on ActiveRecord::Statement(Query).

**Comment by rafaelfranca** [comment]
cc @sgrif

**Comment by sgrif** [comment]
I can't seem to find the original issue for reference, but basically this is the expected behavior. The reason it's not persisting is because the column hasn't been changed, and if we were to mark it as changed that would break the contract of `User.new.changed? == false`. All of this comes back to a quirk in the original design of `serialize`, that says it never returns `nil`, and therefore casts `nil` to an empty array. While I don't like the behavior, it's very hard to change, and the non-persistence of that column going from 4.1 to 4.2 in this case was actually a bug fix, not a regression.

TL;DR You're right, but you should just change that column to allow `NULL`.

**Comment by veera** [comment]
@sgrif ok
